File: Other/Phase 3/Review2/topological-sort/topological-sort.ts
enum Color {
  White = 0, // non-access
  Gray = 1,
  Black = 2, // both itself and its adjacent nodes were accessed
}

export class Graph {
  vertices: number;
  adjacentList: Map<number, number[]>;

  constructor(size: number) {
    this.vertices = size;
    this.adjacentList = new Map();
  }

  addEdge(v: number, w: number) {
    if (!this.adjacentList.has(v)) {
      this.adjacentList.set(v, []);
    }
    this.adjacentList.get(v)!.push(w);
  }

  topologicalSort() {
    const stack: number[] = [];
    const visited: Color[] = new Array(this.vertices).fill(Color.White);

    const recursive = (vertex: number): boolean => {
      visited[vertex] = Color.Gray;

      if (this.adjacentList.has(vertex)) {
        for (const neighbor of this.adjacentList.get(vertex)!) {
          if (visited[neighbor] === Color.Gray) {
            return true;
          }
          if (visited[neighbor] === Color.White && recursive(neighbor)) {
            return true;
          }
        }
      }

      visited[vertex] = Color.Black;
      stack.push(vertex);
      return false;
    };

    for (let i = 0; i < this.vertices; i++) {
      if (visited[i] === Color.White) {
        if (recursive(i)) {
          return null;
        }
      }
    }

    return stack.reverse();
  }
}

// test
const graph1 = new Graph(6);
graph1.addEdge(5, 2);
graph1.addEdge(5, 0);
graph1.addEdge(4, 0);
graph1.addEdge(4, 1);
graph1.addEdge(2, 3);
graph1.addEdge(3, 1);

const result1 = graph1.topologicalSort();
console.log(result1); // [5, 4, 2, 3, 1, 0]

const graph2 = new Graph(6);
graph2.addEdge(5, 2);
graph2.addEdge(5, 0);
graph2.addEdge(4, 0);
graph2.addEdge(4, 1);
graph2.addEdge(2, 3);
graph2.addEdge(3, 1);
graph2.addEdge(1, 2);

const result = graph2.topologicalSort();
console.log(result); // null

File: Miscellaneous/Other/Review/Stage 6/Review 4/Part1/topological-sort/topological-sort.go
package main

import "fmt"

type Color int

const (
	White Color = iota // non-access
	Gray
	Black // both itself and its adjacent nodes were accessed
)

type Graph struct {
	vertices     int
	adjacentList map[int][]int
}

func New(size int) *Graph {
	return &Graph{
		vertices:     size,
		adjacentList: make(map[int][]int, 0),
	}
}

func (graph *Graph) AddEdge(v int, w int) {
	if _, ok := graph.adjacentList[v]; !ok {
		graph.adjacentList[v] = make([]int, 0)
	}
	graph.adjacentList[v] = append(graph.adjacentList[v], w)
}

func (graph *Graph) TopologicalSort() []int {
	stack := make([]int, 0)
	visited := make([]Color, graph.vertices)

	var recursion func(vertex int) bool
	recursion = func(vertex int) bool {
		visited[vertex] = Gray

		if _, ok := graph.adjacentList[vertex]; ok {
			for _, neighbor := range graph.adjacentList[vertex] {
				if visited[neighbor] == Gray {
					return true
				}
				if visited[neighbor] == White && recursion(neighbor) {
					return true
				}
			}
		}

		stack = append(stack, vertex)
		visited[vertex] = Black
		return false
	}

	for i := 0; i < graph.vertices; i++ {
		if visited[i] == White {
			if recursion(i) == true {
				return nil
			}
		}
	}

	for i, j := 0, len(stack)-1; i < j; i, j = i+1, j-1 {
		stack[i], stack[j] = stack[j], stack[i]
	}

	return stack
}

// test
func main() {
	graph1 := New(6)
	graph1.AddEdge(5, 2)
	graph1.AddEdge(5, 0)
	graph1.AddEdge(4, 0)
	graph1.AddEdge(4, 1)
	graph1.AddEdge(2, 3)
	graph1.AddEdge(3, 1)

	result1 := graph1.TopologicalSort()
	fmt.Println(result1) // [5, 4, 2, 3, 1, 0]

	graph2 := New(6)
	graph2.AddEdge(5, 2)
	graph2.AddEdge(5, 0)
	graph2.AddEdge(4, 0)
	graph2.AddEdge(4, 1)
	graph2.AddEdge(2, 3)
	graph2.AddEdge(3, 1)
	graph2.AddEdge(1, 2)

	result := graph2.TopologicalSort()
	fmt.Println(result) // []
}
